docs/design/2024-05-10-extension-authentication-plugin.md

```diff
@@ -0,0 +1,231 @@
+  + Errors returned from this check will cause the connect attempt to fail
++ Allow user defined extra SQL privilege check to be executed before SQL statements are executed
+
+Note that the above plugin SQL privilege checks are only additional to the existing checks done by TiDB. If the existing checks done by TiDB fails, the plugin checks won’t be invoked. This means that the user should already have the corresponding SQL privileges granted to itself before we can trigger the plugin privilege check, otherwise the SQL request should be rejected beforehand since the user is missing the privileges.
+
+Users should be able to develop custom auth plugins which handle the above two checks. For example, if a user develops a plugin called `authentication_my_plugin`, then once the plugin is loaded, the user can run the following SQL statement:
+
+```sql
+CREATE USER 'my_user'@'%' IDENTIFIED WITH 'authentication_my_plugin' AS 'optional_authentication_pwd';
+```
+
+After this, `my_user`’s connection verification and privilege verification will be handled by the plugin in addition to the existing default TiDB checks. Users not identified with the plugin will not go through the plugin authorization and authentication checks.
+
+### Interface
+We propose a set of interfaces for users to implement an auth plugin that are similar to what is offered in MySQL’s authentication plugins. We will modify TiDB’s code to call these interfaces during the authentication/authorization flows. Most of these interfaces will receive parameters that are used in the existing auth flows in TiDB, so that the plugin developer can access as much information as possible.
+
+Following is the interface for an auth plugin:
+```go
+// AuthPlugin is an interface for authentication plugins.
+type AuthPlugin interface {
+    // Name returns the name of the auth plugin. It will be registered as a system variable in TiDB which can be used inside the `CREATE USER ... IDENTIFIED WITH 'plugin_name'` statement.
+    Name() string
+    
+    // PasswordReader is a function to communicate with the client during handshake phase to 
+    // read/write more data to the client-side plugin using authConn. For example, sending extra salt to the client.
+    // If `plugin.PasswordReader()` returns nil, the connection will default to mysql_clear_password plugin which means the password will be passed from the client-side in clear text format. Otherwise, the connection will use `plugin.Name()` as the current plugin
+    // When `plugin.PasswordReader()(ctx, authConn, resp)` is called, it should return the final password received from the client in bytes format. This bytes password will be passed into `plugin.Authenticate` below.
+    // ctx: context passed in from caller
+    // authConn: interface for the plugin to communicate with the client
+    PasswordReader() func(ctx context.Context, authConn conn.AuthConn) ([]byte, error)
```

Comment:
Maybe "tell the client to use a specified plugin proto" and "read the custom password" are two orthogonal functions. Sometimes client may want to keep simple and only use the exist plugins. For example, if we want to introduce a new plugin `caching_sha2_password_enhanced`, it not only checks the passwords from the user as the old way but also does some custom verifications of the SSL certificates or some other checks. We have to implement a client-side plugin because `caching_sha2_password` cannot be used by client in the current design.

How about just give a method: `RequireClientSidePlugin() string` to just indicate the client-side plugin it should use. If it is not set, the plugin in name is used.

`PasswordReader` is not necessary in this way, because the coder can do it in `Authenticate` internally by calling related method of `conn.AuthConn`.

Reply:
I see. I can add the function `RequiredClientSidePlugin() string` which returns the client-side plugin to use.

I did a quick test and I was able to implement the exact same server-side behavior for `caching_sha2_password` by using `authConn` in `Authenticate`. I will remove `PasswordReader` then. Thanks for the suggestion!